[Clang][Driver][SamplePGO] Introduce -fno_sample_profile_use_profi flag for SamplePGO

nilanjana87 opened the pull request:
Introduce -fno_sample_profile_use_profi flag for SamplePGO mode
This is part1 of the set of patches that is needed to [enable -fsample-profile-use-profi flag by default for SamplePGO](https://github.com/llvm/llvm-project/pull/146795).
Discourse conversation: https://discourse.llvm.org/t/turn-on-fsample-profile-use-profi-by-default/86508/2

nilanjana87 committed:
[Clang][Driver][SamplePGO] Enable -fsample-profile-use-profi for SamplePGO in clang

nilanjana87 commented:
> We evaluated profi internally (at Google) last year. Our configuration uses AutoFDO with [flow-sensitive discriminators ](https://lists.llvm.org/pipermail/llvm-dev/2020-November/146694.html) (FS-AFDO). We found slight regressions with this configuration and didn't enable it for our workloads. With FS-AFDO we see performance improvements of ~1%. Can you share your configuration and performance numbers?

Our configuration uses SamplePGO with non-FS discriminators. We experimented with some benchmarks & 2 real-world workloads. Our experiments show better perf gains for 9/13 cases with this flag enabled, with 3 of them showing > 1% extra gain. On the downside of using this flag, we have seen >1% net loss in perf. gain for 3 benchmarks. For the real-world use-cases, we typically see 7-11% perf gain with SamplePGO & ~1% improvement on top of that with this flag enabled.

My argument in favor of enabling this flag by default is that increasing coverage of samples is ideally a good thing. If increased coverage shows reduced performance, then that can be addressed separately by disabling this flag for those particular cases. Empirically too, we see more cases with wins than losses on enabling this by default.

snehasish commented:
Sounds reasonable to me. Can you split this into two changes, first let's commit the one which adds the opt out flag and then the one which makes it the default?

nilanjana87 committed:
part 1: introducing -fno_sample_profile_use_profi flag

Breaking down the patch into 2 parts:
Keeping part1: introducing -fno_sample_profile_use_profi flag
Reverting part2: enabling -fsample_profile_use_profi flag by default

nilanjana87 commented:
> Sounds reasonable to me. Can you split this into two changes, first let's commit the one which adds the opt out flag and then the one which makes it the default?

Updated this patch to reflect part 1 of the split i.e. introducing the opt-out flag.

snehasish commented on clang/test/Driver/pgo-sample-use-profi.c:
```diff
@@ -1,4 +1,19 @@
-/// Test if profi flat is enabled in frontend as user-facing feature.
-// RUN: %clang --target=x86_64 -c -fsample-profile-use-profi -fprofile-sample-use=/dev/null -### %s 2>&1 | FileCheck %s
+/// Ensure that profi flag is enabled by default in frontend for SamplePGO
```
Update this comment.

nilanjana87 replied:
Done

snehasish reviewed (approved):
lgtm

nilanjana87 committed:
nit: Updated comment

nilanjana87 commented:
> Could you do a workaround like this [f92551b](https://github.com/llvm/llvm-project/commit/f92551ba6b5aaef28d886ea6b3c4f9774fa8fadd)? Otherwise, might have to make it unsupported for AIX.

Added workaround in https://github.com/llvm/llvm-project/pull/146996.
